compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -6643,6 +6644,45 @@ public void visit(BLangAnnotAccessExpr annotAccessExpr, AnalyzerData data) {
         data.resultType = this.types.checkType(annotAccessExpr, actualType, data.expType);
     }
 
+    @Override
+    public void visit(BLangNaturalExpression naturalExpression, AnalyzerData data) {
+        BType type = data.expType;
+        SemType expTypeSemType = type.semType();
+        boolean constNaturalExpr = naturalExpression.constExpr;
+
+        if (!constNaturalExpr && !types.isSubtype(symTable.errorType, expTypeSemType)) {
+            dlog.error(naturalExpression.pos, DiagnosticErrorCode.EXPECTED_TYPE_FOR_NATURAL_EXPR_MUST_CONTAIN_ERROR);
```

Comment:
Cant we use the regular incompatible types error here. The main issue is in the error message we say the expected type is incorrect but the position we use for the error msg is the expression pos

Reply:
My concern is, if we consider an assignment, we are saying that the LHS type is incorrect while giving the error on the RHS.

Reply:
It's a commn concern with dependently-typed expressions. I think the current error explains the issue better to the user compared to giving an error to the expected type. The error needs to work with all kinds of places in which a natural expression is used.